Renaming the node role search to warm

vinaykpud opened the pull request:
In this PR we are renaming the existing Node "Search Role" to "Warm Role"

### Description

This is done based on the decision taken as part of the discussion in this thread: https://github.com/opensearch-project/OpenSearch/issues/17422#issuecomment-2712163400

### Related Issues
Related to #15306 
Related to #17422

### Check List
- [x] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created] https://github.com/opensearch-project/documentation-website/issues/9392

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

mch2 commented:
This is a straight rename correct? It does not resolve https://github.com/opensearch-project/OpenSearch/issues/17422 ?

vinaykpud commented:
> This is a straight rename correct? It does not resolve #17422 ?

Yes, Do you want me to add a new issue for this?

andrross commented:
Let's make sure this aligns with what @gbbafna mentioned [here](https://github.com/opensearch-project/OpenSearch/issues/17422#issuecomment-2709489834) before merging.

We'd generally want to do a deprecation cycle before just replacing a term like this. However, it does appear that `search` was a bit of a misnomer and it is not in wide use beyond searchable snapshots, so I'm not totally opposed with doing this to unblock the usage of the `search` role where it might actually make more sense.

gbbafna commented:
@vinaykpud : Can we verify if we upgrade the cluster from 2.x to 3.0 with this change, searchable snapshots continue to work ? There might be few nuances here like changing the cluster-manager first . Want to make sure that rolling restarts works with this change

bugmakerrrrrr commented on server/src/main/java/org/opensearch/cluster/node/DiscoveryNode.java:
```diff
@@ -480,12 +480,12 @@ public boolean isRemoteClusterClient() {
 
     /**
-     * Returns whether the node is dedicated to provide search capability.
+     * Returns whether the node is dedicated to provide searchable snapshot.
```
warm node also for writable warm index, right?

vinaykpud replied:
I tested this, here is the error i got. So its not a writable index now.

```
{
   "error":{
      "root_cause":[
         {
            "type":"cluster_block_exception",
            "reason":"index [products] blocked by: [FORBIDDEN/13/remote index is read-only];"
         }
      ],
      "type":"cluster_block_exception",
      "reason":"index [products] blocked by: [FORBIDDEN/13/remote index is read-only];"
   },
   "status":403
}
```

Also from the docs: https://opensearch.org/docs/latest/tuning-your-cluster/availability-and-recovery/snapshots/searchable_snapshot/ 
>  Any attempt to write to a searchable snapshot index results in an error.

gbbafna replied:
Yes , as of now writable warm indices and searchable snapshots use same role. 

@vinaykpud  : searchable snapshots are not writable index. There is a different API to create those indices .

mch2 replied:
I think the feedback here is in the comment, not any particular API response.

"dedicated to hold warm indices" covers both writeable warm and ss.

vinaykpud replied:
Sure, @bugmakerrrrrr @rayshrey  I was not aware of the writable_warm. Thanks. I will update this.

bugmakerrrrrr commented on server/src/main/java/org/opensearch/cluster/node/DiscoveryNodeRole.java:
```diff
@@ -298,13 +298,13 @@ public Setting<Boolean> legacySetting() {
 
     /**
-     * Represents the role for a search node, which is dedicated to provide search capability.
+     * Represents the role for a warm node, which is dedicated to provide searchable snapshots.
```
here too

gbbafna replied:
add `warm` indices as well .

bugmakerrrrrr commented on server/src/main/java/org/opensearch/common/settings/ClusterSettings.java:
```diff
@@ -709,7 +709,7 @@ public void apply(Settings value, Settings current, Settings previous) {
 
                 // Settings related to Searchable Snapshots
-                Node.NODE_SEARCH_CACHE_SIZE_SETTING,
+                Node.NODE_WARM_CACHE_SIZE_SETTING,
```
this may not need to be renamed, we usually concatenate the settings key as the variable name.

vinaykpud replied:
Should we consider renaming the settings string as well? which I feel more appropriate as we already taken decision to update the node role to warm. 
ie `node.search.cache.size` => `node.warm.cache.size`

This will not impact users because as part of this upgrade they anyway need to update the node role in YML file from `search` to `warm`, so they can update this setting as well. 

Tagging: @andrross @mch2 @gbbafna 

Eg: Old Config
```
      - node.roles=search
      - node.search.cache.size=50gb 
      - 
 ```
new Config:
```
      - node.roles=warm
      - node.warm.cache.size=50gb 
      - 
 ```

bugmakerrrrrr replied:
I was thinking if we have some other caches, for example write cache,  on the warm node in the future, which name should we give to it? IMO, the current name is ok, we can interpret it as the search cache size on the node, or we can rename it to something like `node.warm.search_cache.size`/`searchable_snapshots.cache.size`.

gbbafna replied:
+1, else this will be too confusing for the users.

vinaykpud replied:
Okay, So lets not rename this. I am going to keep the `NODE_SEARCH_CACHE_SIZE_SETTING` and setting as `node.search.cache.size`

vinaykpud commented:
> @vinaykpud : Can we verify if we upgrade the cluster from 2.x to 3.0 with this change, searchable snapshots continue to work ? There might be few nuances here like changing the cluster-manager first . Want to make sure that rolling restarts works with this change

Performed a test to check the cluster upgrade scenario. Using [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) I have setup cluster with 3 cluster manager and 2 data nodes. 
Here is the summary of the steps performed.

1. Created index with 1P and 1R and indexed few documents
2. Registered a S3 repo for snapshots
3. took a snapshot
4. Deleted the index
5. Updated the node role from `data` to `search` by changing the yaml file, restarted the process in both data nodes to make them snapshot search nodes
6. restored searchable snapshot and made sure shards are assigned
7.  created tarball of the OpenSearch with changes in this PR and copied to all the nodes
8. Restarted the OS process with latest binaries first in each cluster managers one by one and cluster was healthy.
9. Then restarted the OS process with latest binaries in each search node one after another.
10. During this process I had `_cat/nodes`, `_cat/shards`, `health` APIs running in loop for monitoring the cluster. 
11. Cluster din't go Red and upgraded to OS 3.0.0
12. Also verified by querying `_search` on the searchable snapshot index and its working.

Let me know if we missed anything here.

mch2 commented:
> > @vinaykpud : Can we verify if we upgrade the cluster from 2.x to 3.0 with this change, searchable snapshots continue to work ? There might be few nuances here like changing the cluster-manager first . Want to make sure that rolling restarts works with this change
> 
> Sure @gbbafna
> 
> Performed a test to check the cluster upgrade scenario. Using [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) I have setup cluster with 3 cluster manager and 2 data nodes. Here is the summary of the steps performed.
> 
> 1. Created index with 1P and 1R and indexed few documents
> 2. Registered a S3 repo for snapshots
> 3. took a snapshot
> 4. Deleted the index
> 5. Updated the node role from `data` to `search` by changing the yaml file, restarted the process in both data nodes to make them snapshot search nodes
> 6. restored searchable snapshot and made sure shards are assigned
> 7. created tarball of the OpenSearch with changes in this PR and copied to all the nodes
> 8. Restarted the OS process with latest binaries first in each cluster managers one by one and cluster was healthy.
> 9. Then restarted the OS process with latest binaries in each search node one after another.
> 10. During this process I had `_cat/nodes`, `_cat/shards`, `health` APIs running in loop for monitoring the cluster.
> 11. Cluster din't go Red and upgraded to OS 3.0.0
> 12. Also verified by querying `_search` on the searchable snapshot index and its working.
> 
> Let me know if we missed anything here.
> 
> cc @mch2 @andrross

So to summarize, you did a successful rolling upgrade from 2.19 to 3.0 but it required cm's to migrate first.  Did you not have to update the routing pool logic for bwc?  I don't see this in the pr. 
```java
    public static RoutingPool getNodePool(DiscoveryNode node) {
        if (node.isWarmNode() || (node.isSearchNode() && (node.getVersion().before(Version.V_3_0_0)))) {
            return REMOTE_CAPABLE;
        }
        return LOCAL_ONLY;
    }
```

vinaykpud commented:
> > > @vinaykpud : Can we verify if we upgrade the cluster from 2.x to 3.0 with this change, searchable snapshots continue to work ? There might be few nuances here like changing the cluster-manager first . Want to make sure that rolling restarts works with this change
> > 
> > 
> > Sure @gbbafna
> > Performed a test to check the cluster upgrade scenario. Using [opensearch-cluster-cdk](https://github.com/opensearch-project/opensearch-cluster-cdk) I have setup cluster with 3 cluster manager and 2 data nodes. Here is the summary of the steps performed.
> > 
> > 1. Created index with 1P and 1R and indexed few documents
> > 2. Registered a S3 repo for snapshots
> > 3. took a snapshot
> > 4. Deleted the index
> > 5. Updated the node role from `data` to `search` by changing the yaml file, restarted the process in both data nodes to make them snapshot search nodes
> > 6. restored searchable snapshot and made sure shards are assigned
> > 7. created tarball of the OpenSearch with changes in this PR and copied to all the nodes
> > 8. Restarted the OS process with latest binaries first in each cluster managers one by one and cluster was healthy.
> > 9. Then restarted the OS process with latest binaries in each search node one after another.
> > 10. During this process I had `_cat/nodes`, `_cat/shards`, `health` APIs running in loop for monitoring the cluster.
> > 11. Cluster din't go Red and upgraded to OS 3.0.0
> > 12. Also verified by querying `_search` on the searchable snapshot index and its working.
> > 
> > Let me know if we missed anything here.
> > cc @mch2 @andrross
> 
> So to summarize, you did a successful rolling upgrade from 2.19 to 3.0 but it required cm's to migrate first. Did you not have to update the routing pool logic for bwc? I don't see this in the pr.
> 
> ```java
>     public static RoutingPool getNodePool(DiscoveryNode node) {
>         if (node.isWarmNode() || (node.isSearchNode() && (node.getVersion().before(Version.V_3_0_0)))) {
>             return REMOTE_CAPABLE;
>         }
>         return LOCAL_ONLY;
>     }
> ```

@mch2 , No, I haven't updated the routing pool logic, it worked without adding any new logic.
But I haven't tried other way ie: First Upgrading Search/Warm Nodes then Upgrading CM's later. We might need to add this logic for that.

mch2 commented:
> First Upgrading Search/Warm Nodes and Upgrading CM's later
@vinaykpud No this logic would be on 3.0 cms not 2.x.  I was thinking this would provide added safety for mixed cluster cases where we need to make alloc decisions and assign ss shards in that mixed state. Though in the normal upgrade case this isn't required.  I think we can go ahead without this.
